pkg/mutation/system.go

```diff
@@ -1,20 +1,173 @@
 package mutation
 
 import (
+	"fmt"
+	"sort"
 	"sync"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
 // System keeps the list of mutations and
 // provides an interface to apply mutations.
 type System struct {
-	// schemaDB schema.DB commented to please the linter, will uncomment in the implementation
-	mutators []Mutator
+	schemaDB        SchemaDB
+	orderedMutators []Mutator
+	mutatorsMap     map[ID]Mutator
 	sync.RWMutex
 }
 
 // NewSystem initializes an empty mutation system
 func NewSystem() *System {
 	return &System{
-		mutators: make([]Mutator, 0),
+		orderedMutators: make([]Mutator, 0),
+		mutatorsMap:     make(map[ID]Mutator),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	current, ok := s.mutatorsMap[m.ID()]
+	if ok && !m.HasDiff(current) {
+		return nil
+	}
+
+	toAdd := m.DeepCopy()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := toAdd.(MutatorWithSchema); ok {
+		err := s.schemaDB.Upsert(withSchema)
+		if err != nil {
+			return errors.Wrapf(err, "Schema upsert failed")
+		}
+	}
+
+	s.mutatorsMap[toAdd.ID()] = toAdd
+
+	i := sort.Search(len(s.orderedMutators), func(i int) bool {
+		return greaterOrEqual(s.orderedMutators[i], toAdd)
+	})
+
+	if i == len(s.orderedMutators) { // Adding to the bottom of the list
+		s.orderedMutators = append(s.orderedMutators, toAdd)
+		return nil
+	}
+
+	found := equal(s.orderedMutators[i], toAdd)
+	if found {
+		s.orderedMutators[i] = toAdd
+		return nil
+	}
+
+	s.orderedMutators = append(s.orderedMutators, nil)
+	copy(s.orderedMutators[i+1:], s.orderedMutators[i:])
+	s.orderedMutators[i] = toAdd
+	return nil
+}
+
+// Mutate applies the mutation in place to the given object
+func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) error {
+	s.RLock()
+	defer s.RUnlock()
+
+	maxIterations := len(s.orderedMutators) + 1
+
+	old := obj.DeepCopy()
```

Comment:
I think we'll need to reset `old` in the outer for-loop (e.g. after line 84)

Reply:
It's getting resetted in line 96, which is basically the same, but I agree it's more readable this other way.
Removing both and adding a single one in the outer cycle.

Reply:
ah, missed line 96, glad the comment was helpful anyway.